### PR TITLE
GET updated data for a Refund immediately after it has been created

### LINKED ISSUE #6704: [Refunds] After refunding, the app always shows "via manual refund" until the Order is refreshed
**Describe the bug**
A clear and concise description of what the bug is. Please be as descriptive as possible; issues lacking detail, or for any other reason than to report a bug, may be closed without action.

**To Reproduce**
1. Go to an order, paid via a payment gateway which supports automatic refunds (e.g. WCPay)
2. Tap `Issue Refund`, and refund the payment to the original payment method
3. On the order details, note that the subtitle incorrectly says "via manual refund"
4. Pull to refresh the order
5. Note that the subtitle now correctly shows the refund payment method

**Screenshots**

https://user-images.githubusercontent.com/2472348/165113846-6423fa8b-dbc4-41d7-b4ab-95e32852846d.mp4

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by iamgabrielma**
Closes: #6704

### Description
When we create a refund via the app, the process involves submitting a POST request to `/wc/v3/orders/$ORDER_ID/refunds&_method=post`. This data is used to map the refund model before we can dispatch it to the store.

One of the properties that we expect from this POST request is [refunded_payment](https://woocommerce.github.io/woocommerce-rest-api-docs/#order-refund-properties). This defines if the payment gateway API is used to generate the refund (or, on the contrary, it is a manual refund), however, the POST response doesn't seem to return this property at the moment, which seems to be a bug with the API as the usage of `refunded_payment` can be seen in the [documentation and examples](https://woocommerce.github.io/woocommerce-rest-api-docs/#create-a-refund).

If we perform a GET request to `/wc/v3/orders/$ORDER_ID/refunds&_method=get ` in order to retrieve an existing refund data, the parameter `refunded_payment` can be found in the response.

This causes that when we create a refund via the app, the [refund summary](https://github.com/woocommerce/woocommerce-ios/blob/46196e64c17beaeac9388ea509c068917d168e8c/WooCommerce/Classes/ViewModels/Order%20Details/OrderPaymentDetailsViewModel.swift#L122) always defaults to "manual payment", and the data doesn't update until we perform a GET request (for example, by refreshing the View). This means the user is seeing wrong information on-screen unless they refresh the View.

This PR attempts to resolve this problem by adding a GET request via the existing action [RefundAction.retrieveRefund()](https://github.com/woocommerce/woocommerce-ios/blob/81ed7b18078c58d3337f53a91d0f09c802bf3c3f/Yosemite/Yosemite/Actions/RefundAction.swift#L9) immediately after the refund has been submitted to be created, on its completion handler.

### Testing instructions
1 - Go to an order that has been paid via a payment gateway that supports automatic refunds (e.g. Stripe, WCPay, ...)
2 - Tap Issue Refund, and refund the payment to the original payment method
3 - On the order details, see that now correctly shows the refund payment method

### Screenshots
Kapture 2022-06-09 at 21 08 08

TODO:
- [x] Open the GH issue with the API and add the link to the code comment
- [ ] Unit Tests?
- [x] Question: Is there a different way to write [this](https://github.com/woocommerce/woocommerce-ios/pull/7069/files#diff-427d3b9e04aaa71070d08f4a840f9c40fa106c8e2247f1799e3098efdd3a3afdR388)? I'm not using the completion handler here, just need to write it because the original method deals with the error at that point.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift [hunk 1]
```diff
@@ -390,6 +398,21 @@ private extension RefundSubmissionUseCase {
         stores.dispatch(action)
         trackCreateRefundRequest()
     }
+
+    /// Retrieves the up-to-date refund data
+    /// - Parameters:
+    ///   - refund: the refund to retrieve details from.
+    ///   - onCompletion: called when the submission completes.
+    func getLatestRefundData(refund: Refund, onCompletion: @escaping (Result<Void, Error>) -> Void) {
```
Comment (toupper) [inline WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift]: I see that this function is just dispatching the `retrieveRefund` action, can we call that one directly without the need to create another function?
Otherwise, if you just want that function to encapsulate the action call it is fine :) In that case, I would:
- Make it private
- Remove the completion parameter as it seems like we do not need it :)

File: WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift [hunk 2]
```diff
@@ -377,8 +377,16 @@ private extension RefundSubmissionUseCase {
...
-        let action = RefundAction.createRefund(siteID: details.order.siteID, orderID: details.order.orderID, refund: refund) { [weak self] _, error  in
+
+        let action = RefundAction.createRefund(siteID: details.order.siteID, orderID: details.order.orderID, refund: refund) { [weak self]
+            refundData, error  in
+
             guard let self = self else { return }
+
+            if let refundData = refundData {
+                // Workaround for https://github.com/woocommerce/woocommerce-ios/issues/6704. It can be removed when the API issue is fixed (Link TBD)
+                self.getLatestRefundData(refund: refundData) { [weak self] _ in }
```
Comment (toupper) [inline WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift]: As explained below, perhaps we do not need the completion block (or even creating a function). But if the completion block remains, we do not need `[weak self]` here, because `self` is nowhere used.

**Comment by toupper** [comment]
Great job @iamgabrielma! Especially here adding a concise and explanatory Pull Request description; really made it much easier to understand what kind of problem are we trying to fix here.

**Comment by joshheald** [comment]
> ### Description
> When we create a refund via the app, the process involves submitting a POST request to `/wc/v3/orders/$ORDER_ID/refunds&_method=post`. This data is used to map the refund model before we can dispatch it to the store.
> ...

Kudos on the clear PR description 👍 

The one improvement I'd suggest is to rearrange the paragraphs, moving the description of the user-facing issue to the front so that it reads as follows:

> When we create a refund via the app, the [refund summary](https://github.com/woocommerce/woocommerce-ios/blob/46196e64c17beaeac9388ea509c068917d168e8c/WooCommerce/Classes/ViewModels/Order%20Details/OrderPaymentDetailsViewModel.swift#L122) always defaults to "manual payment", and the data doesn't update until we perform a GET request (for example, by refreshing the View). This means the user is seeing wrong information on-screen unless they refresh the View.
>
> The refund process involves submitting a POST request to /wc/v3/orders/$ORDER_ID/refunds&_method=post. This data is used to map the refund model before we can dispatch it to the store.
> 
> One of the properties that we expect from this POST request is [refunded_payment](https://woocommerce.github.io/woocommerce-rest-api-docs/#order-refund-properties). This defines if the payment gateway API is used to generate the refund (or, on the contrary, it is a manual refund), however, the POST response doesn't seem to return this property at the moment, which seems to be a bug with the API as the usage of refunded_payment can be seen in the [documentation and examples](https://woocommerce.github.io/woocommerce-rest-api-docs/#create-a-refund).
>
> If we perform a GET request to /wc/v3/orders/$ORDER_ID/refunds&_method=get in order to retrieve an existing refund data, the parameter refunded_payment can be found in the response.
> 
> This PR attempts to resolve this problem by adding a GET request via the existing action [RefundAction.retrieveRefund()](https://github.com/woocommerce/woocommerce-ios/blob/81ed7b18078c58d3337f53a91d0f09c802bf3c3f/Yosemite/Yosemite/Actions/RefundAction.swift#L9) immediately after the refund has been submitted to be created, on its completion handler.


That way, you lay out the issue, then the reason it's happening, and finally the solution used.

The TODO/Questions section is super handy, and so is the GIF.

File: WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift (hunk 1 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift]: I agree with @toupper here...

However, a possible use for the completion would be to delay the dismissing of the overlay view until the GET request had completed, so we could be sure that the UI would correctly describe the problem.

This would probably just be adding an unnecessary delay for the user and making the app feel slower, and we could still avoid incorrectly telling the user that there was a manual refund by displaying `isAutomated == nil` as simply "Refunded", or "Refunded (method unknown)" in the UI.

File: WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift [hunk 3]
```diff
@@ -390,6 +398,21 @@ private extension RefundSubmissionUseCase {
...
         trackCreateRefundRequest()
     }
+
+    /// Retrieves the up-to-date refund data
+    /// - Parameters:
+    ///   - refund: the refund to retrieve details from.
+    ///   - onCompletion: called when the submission completes.
+    func getLatestRefundData(refund: Refund, onCompletion: @escaping (Result<Void, Error>) -> Void) {
+        let action = RefundAction.retrieveRefund(siteID: details.order.siteID, orderID: details.order.orderID, refundID: refund.refundID) {
+                (refundDataFromGETreq, error) in
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift]: `refundDataFromGETreq`

There are 2 hard things in programming: naming, cache invalidation, and off by one errors ;)

I think this is a bit wordy and, and the view layer shouldn't care where the refund data is coming from.

You're also not (really) using it for anything, so just make this
```suggestion
                (_, error) in
```

File: WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift [hunk 4]
```diff
@@ -390,6 +398,21 @@ private extension RefundSubmissionUseCase {
...
+            if let error = error {
+                DDLogError("Error retrieving refund data: \(String(describing: refundDataFromGETreq))\nWith Error: \(error)")
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift]: I said above you're not really using the refund data for anything, and that's because it'll always be nil here. The signature of this action is confusing and you're not helped by the type system at all, so it's not surprising that you expected you'd get both things... but you won't.

If you take a look at the [`RefundStore.retrieveRefund` implementation](https://github.com/woocommerce/woocommerce-ios/blob/39b95821442ebbaff94d53d886fe5230e40381b6/Yosemite/Yosemite/Stores/RefundStore.swift#L72), you'll see that you'll never get both properties here. 

It's a convention in these older responses to return a tuple of optionals, one half of which might have the data, the other the error. We just expect and trust that you'll always get one or the other, never both or neither.

Newer responses are better because we use a `Result<Refund, Error>`, which will either be `.success(refund)` or `.failure(error)`, which means you get the type system helping with your handling of the response.

Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift]:
```suggestion
                DDLogError("Error retrieving refund: \(String(describing: refund))\nWith Error: \(error)")
```

File: WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift [hunk 5]
```diff
@@ -390,6 +398,21 @@ private extension RefundSubmissionUseCase {
...
+                return onCompletion(.failure(error))
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift]: I don't think we would mark the whole refund as a failure in the UI for this... but I may be misunderstanding and I've not tested it. Let's look at this bit when we pair 😊

**Review by joshheald — APPROVED** [review]
A few minor changes, but great job!

File: WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift (hunk 1 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift]: That makes sense. I'll get rid of the completion handler and keep the separate private function, so "submitting the refund" and "updating the refund" are encapsulated in their respective methods.

File: WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift (hunk 3 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift]: Oops! This one flew under the radar in the process of testing -> cleaning things up for the PR 😅 Sorry for that!

File: WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift (hunk 4 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift]: Gotcha! You're correct that my first impression when I saw the syntax for `( something, something_else) in` was that it will return some sort of type with these 2 values. As the error was never triggered during my testing, I always got the data and assumed the error handling would work fine.

File: WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift (hunk 5 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Refund/RefundSubmissionUseCase.swift]: You're right, this was one of the questions I had prepared :D. As we commented during the pairing, I believe we can keep the `DDLogError` for future debugging purposes though.

**Review by toupper — APPROVED** [review]
